tiledb/sm/filter/test/unit_run_filter_pipeline.cc

```diff
@@ -0,0 +1,2010 @@
+      nelts * sizeof(uint64_t));  // First chunk filtered size
+  offset += sizeof(uint32_t);
+  CHECK(
+      tile.filtered_buffer().value_at_as<uint32_t>(offset) ==
+      sizeof(uint64_t));  // First chunk metadata size
+  offset += sizeof(uint32_t);
+
+  // Checksum
+  CHECK(
+      tile.filtered_buffer().value_at_as<uint64_t>(offset) ==
+      expected_checksum);
+  offset += sizeof(uint64_t);
+
+  // Check all elements incremented.
+  for (uint64_t i = 0; i < nelts; i++) {
+    CHECK(tile.filtered_buffer().value_at_as<uint64_t>(offset) == (i + 4));
+    offset += sizeof(uint64_t);
+  }
+
+  auto unfiltered_tile = create_tile_for_unfiltering(nelts, tile);
+  run_reverse(config, tp, unfiltered_tile, pipeline);
+
+  for (uint64_t i = 0; i < nelts; i++) {
+    uint64_t elt = 0;
+    CHECK_NOTHROW(
+        unfiltered_tile.read(&elt, i * sizeof(uint64_t), sizeof(uint64_t)));
+    CHECK(elt == i);
+  }
+}
+
+TEST_CASE("Filter: Test random pipeline", "[filter][random]") {
+  tiledb::sm::Config config;
+
+  const uint64_t nelts = 100;
+
+  EncryptionKey encryption_key;
+  REQUIRE(encryption_key
+              .set_key(
+                  EncryptionType::AES_256_GCM,
+                  "abcdefghijklmnopqrstuvwxyz012345",
+                  32)
+              .ok());
+
+  // List of potential filters to use. All of these filters can occur anywhere
+  // in the pipeline.
+  std::vector<std::function<tiledb::sm::Filter*(void)>> constructors = {
+      []() { return tdb_new(Add1InPlace, Datatype::UINT64); },
+      []() { return tdb_new(Add1OutOfPlace, Datatype::UINT64); },
+      []() { return tdb_new(Add1IncludingMetadataFilter, Datatype::UINT64); },
+      []() { return tdb_new(BitWidthReductionFilter, Datatype::UINT64); },
+      []() { return tdb_new(BitshuffleFilter, Datatype::UINT64); },
+      []() { return tdb_new(ByteshuffleFilter, Datatype::UINT64); },
+      []() {
+        return tdb_new(
+            CompressionFilter,
+            tiledb::sm::Compressor::BZIP2,
+            -1,
+            Datatype::UINT64);
+      },
+      []() { return tdb_new(PseudoChecksumFilter, Datatype::UINT64); },
+      []() { return tdb_new(ChecksumMD5Filter, Datatype::UINT64); },
+      []() { return tdb_new(ChecksumSHA256Filter, Datatype::UINT64); },
+      [&encryption_key]() {
+        return tdb_new(
+            EncryptionAES256GCMFilter, encryption_key, Datatype::UINT64);
+      },
+  };
+
+  // List of potential filters that must occur at the beginning of the pipeline.
+  std::vector<std::function<tiledb::sm::Filter*(void)>> constructors_first = {
+      // Pos-delta would (correctly) return error after e.g. compression.
+      []() { return tdb_new(PositiveDeltaFilter, Datatype::UINT64); }};
+
+  ThreadPool tp(4);
+  for (int i = 0; i < 100; i++) {
+    auto tile = make_increasing_tile(nelts);
+
+    // Construct a random pipeline
+    FilterPipeline pipeline;
+    const unsigned max_num_filters = 6;
+    std::random_device rd;
```

Comment:
@bekadavis9 Does this sounds like a good place to use your RNG? Could you point @jp-dark to it?

Reply:
I added a shortcut story for this: https://app.shortcut.com/tiledb-inc/story/38297